LiteCore/Query/QueryParser.cc

```diff
@@ -1132,6 +1164,102 @@ namespace litecore {
+               }
+            }
+            return ret;
+        }
+
+        inline bool isMetaArgForValue(const string& arg) {
+            return arg.find(kMetaCryptic) == 0;
+        }
+    }
+
+    // Handles ["meta", dbAlias_optional]
+    void QueryParser::metaOp(slice op, Array::iterator& operands) {
+        // Pre-conditions: op == "meta"
+        //                 operands.size() == 0 || operands[0]->type() == kString (dbAlias)
+        string arg;
+        if (operands.count() > 0 && operands[0]->type() == kString) {
+            arg = operands[0]->asString();
+        }
+
+        if (isMetaArgForValue(arg)) {
+            auto it = find_if(&metaKeys[0], &metaKeys[0] + kMetaKeysCount, [&arg](auto x) {
+                return arg.find(x[1]) == 0;
+            });
+            size_t i = it - &metaKeys[0];
+            require(i < kMetaKeysCount, "Unexpected argument of \"meta\": '%s'", arg.c_str());
+            string metaKey = metaKeys[i][0];
+            size_t prefixLen = strlen(metaKeys[i][1]);
+            string dbAlias = arg.substr(prefixLen, arg.length() - prefixLen);
+            writeMetaPropertyGetter(metaKey, dbAlias, false);
+            
+            return;
+        } // if (isMetaValue) {
+        
+        Path path {slice(arg+".id")};
+        const auto& dbIter = verifyDbAlias(path);
+        require(arg.empty() || arg == dbIter->first,
+                "database alias '%s' does not match a declared 'AS' alias", arg.c_str());
+        parseNode(buildMetaDict(dbIter->first));
+    }
+
+    void QueryParser::writeMetaPropertyGetter(const string& metaKey, const string& dbAlias, bool checkDbAlias) {
+        string tablePrefix;
+        if (!checkDbAlias) {
+            if (!dbAlias.empty()) {
+                tablePrefix = quoteTableName(dbAlias) + ".";
+            }
+        } else {
+            Path path {slice(dbAlias+".id")};
+            const auto& dbIter = verifyDbAlias(path);
+            require(dbAlias.empty() || dbAlias == dbIter->first,
+                    "database alias '%s' does not match a declared 'AS' alias", dbAlias.c_str());
+            if (!dbIter->first.empty()) {
+                tablePrefix = quoteTableName(dbIter->first) + ".";
+            }
+        }
+  
+        if (metaKey == "id") {
+            writeMetaProperty(kValueFnName, tablePrefix, "key");
+        } else if (metaKey == "deleted") {
+            writeDeletionTest(dbAlias, true);
+            _checkedDeleted = true;     // note that the query has tested _deleted
+        } else if (metaKey == "revisionID") {
+            _sql << kVersionFnName << "(" << tablePrefix << "version" << ")";
+        } else if (metaKey == "sequence") {
+            writeMetaProperty(kValueFnName, tablePrefix, metaKey.c_str());
+            _checkedExpiration = true;
+        } else if (metaKey == "expiration") {
+            writeMetaProperty(kValueFnName, tablePrefix, metaKey.c_str());
+        }
+    }
```

Comment:
What happens if `metaKey` isn't any of the above legal values? Shouldn't it call `fail`?

Reply:
I meant to let it go through to the main function, where it would fail, as trying to extract property from a dictionary in that the property does not exit: meta().not_a_mete_key.

Reply:
In fact, this is only a optimization measure. "xyz" in meta().xyz does not match, we would compute meta() into a dictionary. Otherwise, the meta() won't be expanded.